nvfuserex has problems taking getitem.

getitem_nvfuser tests failed for a similar reason to https://github.com/Lightning-AI/lightning-thunder/blob/54bb6146ff757905925f8d9ea2197870c4971011/thunder/tests/opinfos.py#L3113-L3115. I can again create a wrapper so `slice` objects don't get passed to FusionDefinitionWrapper. But I'd love to hear thoughts from @jjsjann123 and @kevinstephano who may have a better fix.

_Originally posted by @wujingyue in https://github.com/Lightning-AI/lightning-thunder/issues/191#issuecomment-2119408081_
            

cc @tfogal

To reproduce this, check out branch `wjy/bookend` and run `pytest thunder/tests/test_ops.py -k test_core_vs_torch_consistency_getitem_nvfuser_cuda_float32 -s`. 

The branch disables bookend (a workaround that we want to get rid of) so slice prims are given to nvfuser, adds some debug prints and highlights the failing test. 

The immediate error is 
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
thunder/tests/test_ops.py:76: in test_core_vs_torch_consistency
    result = run_snippet(
thunder/tests/framework.py:516: in run_snippet
    raise ex
thunder/tests/framework.py:508: in run_snippet
    snippet(*args, **kwargs)
thunder/tests/test_ops.py:37: in snippet_torch_consistency
    thunder_result = op(*sample.args, **sample.kwargs)
thunder/__init__.py:620: in fn_
    result = cache_entry.computation_fn(*inps)
/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py:115: in decorate_context
    return func(*args, **kwargs)
/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py:28: in decorate_autocast
    return func(*args, **kwargs)
/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py:28: in decorate_autocast
    return func(*args, **kwargs)
thunder.computation_1:10: in computation
    [t0] = nvFusion0(a, b)
thunder/executors/nvfuserex_impl.py:401: in __call__
    fd = self.get_fd(to_descriptors(args))
thunder/executors/nvfuserex_impl.py:384: in to_descriptors
    return tuple(to_descriptor(arg) for arg in args)
thunder/executors/nvfuserex_impl.py:384: in <genexpr>
    return tuple(to_descriptor(arg) for arg in args)
thunder/executors/nvfuserex_impl.py:372: in to_descriptor
    numbertype, _ = check_same_dtype(*arg)
thunder/core/utils.py:220: in check_same_dtype
    typ = to_dtype(a, true_dtype=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = slice(1, 3, 1)

    def to_dtype(x: Any, /, *, true_dtype: bool = False) -> None | dtype:
        """Exctracts a dtype from an object or class."""

        if x is None or isinstance(x, dtype):
            return x
        if isinstance(x, np.ndarray):
            return _numpy_to_thunder_dtype_map[x.dtype]
        if isinstance(x, np.dtype):
            return _numpy_to_thunder_dtype_map[x.type]
        if isinstance(x, torch.Tensor):
            return _torch_to_thunder_dtype_map[x.dtype]
        if isinstance(x, torch.dtype):
            return _torch_to_thunder_dtype_map[x]
        if isinstance(x, TensorProxyInterface):
            if true_dtype:
                return x.true_dtype
            return x.dtype
        if isinstance(x, dtype):
            return x
        if isinstance(x, NumberProxyInterface):
            return x.python_type
        if isinstance(x, Number):
            return _numberclass_to_numbertype(type(x))
        if isinstance(x, type) and issubclass(x, Number):
            return _numberclass_to_numbertype(x)

>       raise ValueError(f"Trying to extract a dtype from object {x} with unknown type {type(x)}!")
E       ValueError: Trying to extract a dtype from object slice(1, 3, 1) with unknown type <class 'slice'>!

thunder/core/dtypes.py:300: ValueError
```
This is triggered by to_descriptors which tries to compute a cache key for the function signature and fails on `slice`. 

However, even with that check removed, 
```
diff --git a/thunder/executors/nvfuserex_impl.py b/thunder/executors/nvfuserex_impl.py
index bc0708ef..e30bf428 100644
--- a/thunder/executors/nvfuserex_impl.py
+++ b/thunder/executors/nvfuserex_impl.py
@@ -368,13 +368,13 @@ def to_descriptors(args) -> tuple:
         if isinstance(arg, Number):
             return type(arg)
         elif isinstance(arg, tuple):
-            if len(arg) != 0:
-                numbertype, _ = check_same_dtype(*arg)
-                check(
-                    numbertype == int,
-                    lambda: f"tuple in nvfuser only supports int, but found {numbertype}",
-                    exception_type=AssertionError,
-                )
+            # if len(arg) != 0:
+            #     numbertype, _ = check_same_dtype(*arg)
+            #     check(
+            #         numbertype == int,
+            #         lambda: f"tuple in nvfuser only supports int, but found {numbertype}",
+            #         exception_type=AssertionError,
+            #     )
             return type(arg)
         elif isinstance(arg, torch.Tensor):
             return (*get_tensor_descriptor(arg), arg.dtype)
```
nvFuser still fails with the following RuntimeError: 
```
RuntimeError: Tracer cannot infer type of slice(1, 3, 1)
:Only tensors and (possibly nested) tuples of tensors, lists, or dictsare supported as inputs or outputs of traced functions, but instead got value of type slice.
```

--- a/thunder/core/transform_common.py
+++ b/thunder/core/transform_common.py
@@ -86,13 +86,16 @@ def _inplace_copy_sanity_check(extrace: Trace):
 #   that only produce non-proxy objects
 # NOTE needed_proxies is an in/out argument, it takes an initial set of Variables you want to keep, and return
 #   all the needed proxies of the input trace
-def dce(trace: Trace, needed_proxies: set[Variable] = set()) -> Trace:
+def dce(trace: Trace, needed_proxies=None) -> Trace:
     start_time_ns = time.time_ns()
 
     producer_map: ProxyDict = producers(trace)
 
     flat_trace_outputs, _ = tree_flatten(trace.output)
-    needed_proxies.update(tuple(variableify(x) for x in flat_trace_outputs if isinstance(x, Proxy)))
+    if needed_proxies is None:
+        needed_proxies: set[Variable] = set(tuple(variableify(x) for x in flat_trace_outputs if isinstance(x, Proxy)))
+    else:
+        needed_proxies.update(tuple(variableify(x) for x in flat_trace_outputs if isinstance(x, Proxy)))
     dced = []
 
     bsym: BoundSymbol

--- a/thunder/core/transform_common.py
+++ b/thunder/core/transform_common.py
@@ -84,13 +84,15 @@ def _inplace_copy_sanity_check(extrace: Trace):
 # Runs a Dead Code Elimination (DCE) pass
 # NOTE Today we are only interested in computations that produce proxies, so this will eliminate operations
 #   that only produce non-proxy objects
-def dce(trace: Trace) -> Trace:
+# NOTE needed_proxies is an in/out argument, it takes an initial set of Variables you want to keep, and return
+#   all the needed proxies of the input trace
+def dce(trace: Trace, needed_proxies: set[Variable] = set()) -> Trace:
     start_time_ns = time.time_ns()
 
     producer_map: ProxyDict = producers(trace)
 
     flat_trace_outputs, _ = tree_flatten(trace.output)
-    needed_proxies: set[Variable] = set(tuple(variableify(x) for x in flat_trace_outputs if isinstance(x, Proxy)))
+    needed_proxies.update(tuple(variableify(x) for x in flat_trace_outputs if isinstance(x, Proxy)))
     dced = []
 
     bsym: BoundSymbol

--- a/thunder/core/transform_common.py
+++ b/thunder/core/transform_common.py
@@ -86,7 +86,7 @@ def _inplace_copy_sanity_check(extrace: Trace):
 #   that only produce non-proxy objects
 # NOTE needed_proxies is an in/out argument, it takes an initial set of Variables you want to keep, and return
 #   all the needed proxies of the input trace
-def dce(trace: Trace, needed_proxies=None) -> Trace:
+def dce(trace: Trace, needed_proxies: None | set[Variable] =None) -> Trace:
     start_time_ns = time.time_ns()
 
     producer_map: ProxyDict = producers(trace)

--- a/thunder/core/transform_common.py
+++ b/thunder/core/transform_common.py
@@ -86,7 +86,7 @@ def _inplace_copy_sanity_check(extrace: Trace):
 #   that only produce non-proxy objects
 # NOTE needed_proxies is an in/out argument, it takes an initial set of Variables you want to keep, and return
 #   all the needed proxies of the input trace
-def dce(trace: Trace, needed_proxies: None | set[Variable] =None) -> Trace:
+def dce(trace: Trace, needed_proxies: None | set[Variable] = None) -> Trace:
     start_time_ns = time.time_ns()
 
     producer_map: ProxyDict = producers(trace)

--- a/thunder/executors/nvfuserex_impl.py
+++ b/thunder/executors/nvfuserex_impl.py
@@ -619,11 +619,14 @@ class nvFuserExecutor(FusionExecutor):
 
         return False
 
-    def _dce_bsyms(self, output, bsyms: list[BoundSymbol]) -> list[BoundSymbol]:
+    def _dce_bsyms(self, input_list, output, bsyms: list[BoundSymbol]) -> list[BoundSymbol]:
         trace = TraceCtx(None)
         trace.bound_symbols = bsyms
         bsyms.append(prims.python_return.bind(output, output=()))
-        trace = dce(trace)
+        needed_proxies: set[Variable] = set()
+        trace = dce(trace, needed_proxies)
+        # update the input_list by removing the unused inputs
+        input_list[:] = [x for x in input_list if variableify(x) in needed_proxies]
         return list(filter(lambda x: x.sym != prims.python_return, trace.bound_symbols))
 
     def fuse(self, region: Region, fusion_counter: int) -> BoundSymbol:
@@ -636,8 +639,7 @@ class nvFuserExecutor(FusionExecutor):
         flattened_bsyms: list[BoundSymbol] = []
         for bsym in region.bound_symbols:
             flattened_bsyms.extend(self.flatten(bsym))
-
-        flattened_bsyms = self._dce_bsyms(sorted_unique_outputs, flattened_bsyms)
+        flattened_bsyms = self._dce_bsyms(sorted_unique_inputs, sorted_unique_outputs, flattened_bsyms)
 
         fusion_name = f"nvFusion{fusion_counter}"
         annotation = f"{fusion_name}: ({', '.join(bsym.sym.name for bsym in flattened_bsyms)})"

--- a/thunder/executors/nvfuserex_impl.py
+++ b/thunder/executors/nvfuserex_impl.py
@@ -639,6 +639,7 @@ class nvFuserExecutor(FusionExecutor):
         flattened_bsyms: list[BoundSymbol] = []
         for bsym in region.bound_symbols:
             flattened_bsyms.extend(self.flatten(bsym))
+
         flattened_bsyms = self._dce_bsyms(sorted_unique_inputs, sorted_unique_outputs, flattened_bsyms)
 
         fusion_name = f"nvFusion{fusion_counter}"

--- a/thunder/tests/test_nvfuser.py
+++ b/thunder/tests/test_nvfuser.py
@@ -921,3 +921,29 @@ def test_matmul(executor, device: str, dtype: dtypes.dtype):
 
         assert len(fusions) == 1
         torch.testing.assert_close(out, torch.matmul(*sample.args))
+
+
+@instantiate(
+    dtypes=NOTHING,
+    executors=(nvFuserExecutor,),
+)
+def test_rm_unused_inputs_of_nvfusion(executor, device, _):
+    import operator
+
+    def foo(t, ab):
+        return operator.getitem(t, ab)
+
+    t = make_tensor(5, 3, device=device, dtype=torch.float32)
+    ab = (slice(3, 1), slice(1, 2))
+    # enable bookend would remove the error and let you look at the trace without fusion.
+    jfoo = thunder.functional.jit(
+        foo,
+        interpretation="python interpreter",
+        cache="no caching",
+        disable_torch_autograd=True,
+        nv_enable_bookend=False,
+    )
+    out = jfoo(t, ab)
+    out_ref = foo(t, ab)
+
+    assert out.equal(out_ref)

